network/transport/v2/handlers.go

```diff
@@ -151,10 +158,114 @@ func (p protocol) handleGossip(peer transport.Peer, msg *Gossip) error {
 	if len(refs) > 0 {
 		// TODO swap for trace logging
 		log.Logger().Infof("received %d new transaction references via Gossip", len(refs))
+		p.sendTransactionListQuery(peer.ID, refs)
 	}
 	p.gManager.GossipReceived(peer.ID, refs...)
 	// TODO compare xor
 	// TODO send new message
 
 	return nil
 }
+
+func (p *protocol) handleTransactionList(peer transport.Peer, envelope *Envelope_TransactionList) error {
+	msg := envelope.TransactionList
+	cid, err := parseConversationID(msg.ConversationID)
+	if err != nil {
+		return fmt.Errorf("invalid conversationID: %w", err)
+	}
+	// TODO convert to trace logging
+	log.Logger().Infof("handling handleTransactionList from peer (peer=%s, conversationID=%s)", peer.ID.String(), cid.String())
+
+	// check if response matches earlier request
+	if err = p.cMan.check(envelope); err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+	for _, tx := range msg.Transactions {
+		transactionRef := hash.FromSlice(tx.Hash)
+		transaction, err := dag.ParseTransaction(tx.Data)
+		if err != nil {
+			return fmt.Errorf("received transaction is invalid (peer=%s, ref=%s): %w", peer, transactionRef, err)
+		}
+
+		present, err := p.state.IsPresent(ctx, transactionRef)
+		if !present {
+			// TODO does this always trigger fetching missing payloads? (through observer on DAG) Prolly not for v2
+			if err = p.state.Add(ctx, transaction, tx.Payload); err != nil {
```

Comment:
Who checks whether `tx.PayloadHash == hash(tx.Payload)`?

Reply:
There's a TODO for that: state.go#93. Will implement.

Reply:
Added to State.Add